From 4a338e3634f5ad22c78340dfee2b32d90993eea4 Mon Sep 17 00:00:00 2001
From: Hauke Mehrtens <hauke.mehrtens@intel.com>
Date: Wed, 23 Sep 2020 17:16:54 +0200
Subject: [PATCH] PONRTSYS-8045: Fix array out of bound access

The intel,extra-subif was increased to 4 for both ports on the SFU in
the dts file. This does not fit into the eth_dev array any more.
Increase it to 8 and add an additional check which prevents such
overflows.
---
 drivers/net/ethernet/lantiq/ltq_eth_drv_xrx500.c | 7 ++++++-
 1 file changed, 6 insertions(+), 1 deletion(-)

diff --git a/drivers/net/ethernet/lantiq/ltq_eth_drv_xrx500.c b/drivers/net/ethernet/lantiq/ltq_eth_drv_xrx500.c
index f3ec71ccc938..67f0f391c1f5 100644
--- a/drivers/net/ethernet/lantiq/ltq_eth_drv_xrx500.c
+++ b/drivers/net/ethernet/lantiq/ltq_eth_drv_xrx500.c
@@ -60,7 +60,7 @@
 
 #define LTQ_ETH_NUM_INTERRUPTS 5
 
-#define NUM_IF 6
+#define NUM_IF 8
 #define PON_IP_HOST_SUBIF_IDX	3
 
 /* Init of the network device */
@@ -2212,6 +2212,11 @@ static int xrx500_of_iface(struct xrx500_hw *hw, struct device_node *iface,
 			priv->extra_subif = extra_subif_param;
 			priv->start = dev_num;
 			priv->end = dev_num + priv->extra_subif;
+			if (priv->end > NUM_IF) {
+				pr_err("%s : Too many sub devices requested: %i\n",
+				       name, priv->end);
+				return -EINVAL;
+			}
 			dev_num += priv->extra_subif;
 
 			/* Traffic between diff subif */
